File: Python/213.House Robber II.py
#抢劫，依然是相邻两个家庭会触发警报，但是这个街区是圆形的，第一家和最后一家相邻。求最大收益
#由于第一家和最后一家不能同时出现，分别去掉第一家和最后一家各求一次，比较即可
class Solution(object):
    def rob(self, nums):
        l = len(nums)
        if l == 0:    return 0
        if l == 1:    return nums[0]
        dp = [0] * l
        #去掉第一个
        dp[0], dp[1] = 0, nums[1]
        for i in range(2, l):
            dp[i] = max(dp[i - 2] + nums[i], dp[i - 1])
        answer = dp[l - 1]
        #去掉最后一个
        dp[0], dp[1] = nums[0], max(nums[0], nums[1])
        for i in range(2, l-1):
            dp[i] = max(dp[i - 2] + nums[i], dp[i - 1])
        return max(dp[l - 2], answer)

s = Solution()
print(s.rob([3,1,2,7,9,13,11]))

File: C++/Dynamic Programming/213.House Robber II.cpp
/*
抢劫一个环形社区，相邻的房间被抢会触发警报。
*/
/*
思路：先计算第1到n-1个房子的最大，然后计算第2到n个的最大，比较两者返回最大值。
可以使用两个临时变量存储前前一个和前一个。不用使用整个数组。
 */
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
	int rob(vector<int>& nums) {
		int n = nums.size();
		if(n ==0 )	return 0;
		else if(n==1)	return nums[0];

        int pre = 0, cur = 0 , first;
        for(int i=0; i+1 < n; ++i)
        {
        	int tmp = max(cur,pre+nums[i]);
        	pre = cur;
        	cur = tmp;
        }
        first = cur, pre = 0, cur = 0;
        for(int i=1; i < n; ++i)
        {
        	int tmp = max(cur,pre+nums[i]);
        	pre = cur;
        	cur = tmp;
        }
        return max(first,cur);
    }
};
int main()
{
	Solution s;
	vector<int> nums{2,3,4,4};
	cout<<s.rob(nums)<<endl;
}

